protozfitsreader/__init__.py

```diff
@@ -1,262 +1,313 @@
+]
+
+PATCH_ID_OUTPUT = [
+    204, 216, 229, 241, 254, 266, 279, 291, 304, 316, 329,
+    341, 180, 192, 205, 217, 230, 242, 255, 267, 280, 292, 305, 317, 156, 168,
+    181, 193, 206, 218, 231, 243, 256, 268, 281, 293, 132, 144, 157, 169, 182,
+    194, 207, 219, 232, 244, 257, 269, 108, 120, 133, 145, 158, 170, 183, 195,
+    208, 220, 233, 245, 84, 96, 109, 121, 134, 146, 159, 171, 184, 196, 209,
+    221, 60, 72, 85, 97, 110, 122, 135, 147, 160, 172, 185, 197, 40, 50, 61,
+    73, 86, 98, 111, 123, 136, 148, 161, 173, 24, 32, 41, 51, 62, 74, 87, 99,
+    112, 124, 137, 149, 12, 18, 25, 33, 42, 52, 63, 75, 88, 100, 113, 125, 4,
+    8, 13, 19, 26, 34, 43, 53, 64, 76, 89, 101, 0, 2, 5, 9, 14, 20, 27, 35,
+    44, 54, 65, 77, 228, 239, 251, 262, 274, 285, 297, 308, 320, 331, 343,
+    354, 240, 252, 263, 275, 286, 298, 309, 321, 332, 344, 355, 366, 253, 264,
+    276, 287, 299, 310, 322, 333, 345, 356, 367, 377, 265, 277, 288, 300, 311,
+    323, 334, 346, 357, 368, 378, 387, 278, 289, 301, 312, 324, 335, 347, 358,
+    369, 379, 388, 396, 290, 302, 313, 325, 336, 348, 359, 370, 380, 389, 397,
+    404, 303, 314, 326, 337, 349, 360, 371, 381, 390, 398, 405, 411, 315, 327,
+    338, 350, 361, 372, 382, 391, 399, 406, 412, 417, 328, 339, 351, 362, 373,
+    383, 392, 400, 407, 413, 418, 422, 340, 352, 363, 374, 384, 393, 401, 408,
+    414, 419, 423, 426, 353, 364, 375, 385, 394, 402, 409, 415, 420, 424, 427,
+    429, 365, 376, 386, 395, 403, 410, 416, 421, 425, 428, 430, 431, 215, 191,
+    167, 143, 119, 95, 71, 49, 31, 17, 7, 1, 227, 203, 179, 155, 131, 107, 83,
+    59, 39, 23, 11, 3, 238, 214, 190, 166, 142, 118, 94, 70, 48, 30, 16, 6,
+    250, 226, 202, 178, 154, 130, 106, 82, 58, 38, 22, 10, 261, 237, 213, 189,
+    165, 141, 117, 93, 69, 47, 29, 15, 273, 249, 225, 201, 177, 153, 129, 105,
+    81, 57, 37, 21, 284, 260, 236, 212, 188, 164, 140, 116, 92, 68, 46, 28,
+    296, 272, 248, 224, 200, 176, 152, 128, 104, 80, 56, 36, 307, 283, 259,
+    235, 211, 187, 163, 139, 115, 91, 67, 45, 319, 295, 271, 247, 223, 199,
+    175, 151, 127, 103, 79, 55, 330, 306, 282, 258, 234, 210, 186, 162, 138,
+    114, 90, 66, 342, 318, 294, 270, 246, 222, 198, 174, 150, 126, 102, 78
+]
+
+
+class ZFile:
 
-
-class ZFile(object):
     def __init__(self, fname):
+        if not isfile(fname):
+            raise FileNotFoundError(fname)
         self.fname = fname
+        self.eventnumber = 0
+        self.is_events_table_open = False
+        self.__open_events()
+        self.numrows = rawzfitsreader.getNumRows()
+        self.run_id = 0
 
-    def _read_file(self):
-        if self.ttype not in ["RunHeader", "Events", "RunTails"]:
-            print("Error: Table type not RunHeader, Events or RunTails")
+    def __next__(self):
+        if self.eventnumber < self.numrows:
+            self.__open_events()
+            event = L0_pb2.CameraEvent()
+            event.ParseFromString(rawzfitsreader.readEvent())
+            self.eventnumber += 1
```

Comment:
wouldn't it be better if we access eventnumber from event."something" rather than incrementing by 1? This prevents that numbering from the zfits file and the protozfitsreader differ.

Reply:
Please review again the code as I found it October 2017. this code is behaving exactly like that (unless I made a mistake).

So either something in October 2017 was already wrong .. then we should fix it, or it was all fine in October 2017, then it is still fine (unless I made a mistake)

Whatever the answer is, we should not just change the reader, because "we feel like it is a good idea". We should first write a test, that shows (new users and fellow developers) how the reader should behave, and then we modify the reader, so it behaves that way.

Reply:
Seems that we did not incremented by 1 : 
https://github.com/cta-sst-1m/digicampipe/blob/2df1440705328e1a9886018d59959235eb3be454/digicampipe/io/protozfitsreader.py#L231-L249

```python
        while i < numrows:
            self.read_event()
            # Hook to deal with file with no header (2)
            if hasattr(self, 'numrows'):
                numrows = self.numrows
            # End - Hook to deal with file with no header (2)


            # Hook to deal with file with no header (3)
            try:
                run_id = self.get_run_id()
                event_id = self.get_event_id()
            except:
                run_id = 0
                event_id = self.eventnumber
                # print('No header in this file, run_id set to 0 and event_id set to event_number')
            # Hook to deal with file with no header (3)


            yield run_id, event_id
            i += 1
```

Reply:
Also see : 

https://github.com/cta-sst-1m/digicampipe/blob/2df1440705328e1a9886018d59959235eb3be454/digicampipe/io/protozfitsreader.py#L198-L217